test/framework/control_plane.go

```diff
@@ -34,84 +34,191 @@ import (
-	Nodes             []Node
-	MachineDeployment MachineDeployment
-	RelatedResources  []runtime.Object
-	CreateTimeout     time.Duration
-	DeleteTimeout     time.Duration
+// Interfaces to scope down client.Client
 
-	ControlPlane    *controlplanev1.KubeadmControlPlane
-	MachineTemplate runtime.Object
+// Getter can get resources.
+type Getter interface {
+	Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error
+}
+
+// Creator can creates resources.
+type Creator interface {
+	Create(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) error
+}
+
+// Lister can lists resources.
+type Lister interface {
+	List(ctx context.Context, list runtime.Object, opts ...client.ListOption) error
+}
+
+// Deleter can delete resources.
+type Deleter interface {
+	Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error
+}
+
+// ListDeleter combines interfaces.
+type ListerDeleter interface {
+	Lister
+	Deleter
+}
+
+// Durations allow the customization of a test timeout and polling interval.
+// They can be customized for each test and/or the default can be changed by updating the package globals.
+type Durations struct {
+	Timeout      time.Duration
+	PollInterval time.Duration
 }
 
-// SetDefaults defaults the struct fields if necessary.
-func (input *ControlplaneClusterInput) SetDefaults() {
-	if input.CreateTimeout == 0 {
-		input.CreateTimeout = 10 * time.Minute
+// SetDefaults is a convenience for library consumers so they never have to specify values for this type.
+func (d *Durations) SetDefaults() {
+	if d.Timeout == 0 {
+		d.Timeout = DefaultTimeout
 	}
+	if d.PollInterval == 0 {
+		d.PollInterval = DefaultPollInterval
+	}
+}
 
-	if input.DeleteTimeout == 0 {
-		input.DeleteTimeout = 5 * time.Minute
+// CreateRelatedResourcesInput is the input type for CreateRelatedResources.
+type CreateRelatedResourcesInput struct {
+	Creator          Creator
+	RelatedResources []runtime.Object
+	Durations
+}
+
+// CreateRelatedResources is used to create runtime.Objects.
+func CreateRelatedResources(ctx context.Context, input CreateRelatedResourcesInput) {
+	input.SetDefaults()
+
+	By("creating related resources")
+	for i := range input.RelatedResources {
+		obj := input.RelatedResources[i]
+		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
+		Eventually(func() error {
+			return input.Creator.Create(ctx, obj)
+		}, input.Timeout, input.PollInterval).Should(BeNil())
 	}
 }
 
-// ControlPlaneCluster creates an n node control plane cluster.
-// Assertions:
-//  * The number of nodes in the created cluster will equal the number
-//    of control plane nodes plus the number of replicas in the machine
-//    deployment.
-func (input *ControlplaneClusterInput) ControlPlaneCluster() {
-	ctx := context.Background()
-	Expect(input.Management).ToNot(BeNil())
+// CreateClusterInput is the input for CreateCluster.
+type CreateClusterInput struct {
+	Creator      Creator
```

Comment:
With all of these input structs containing an item for clients, I'm wondering if it would make sense to move the client configuration into a struct and turn the helpers into methods. 

```go
h := NewControlPlaneHelper(myClient)
h.CreateCluster(ctx, ...)
h.CreateKubeadmControlPlane(ctx, ...)
```

Reply:
I like this idea but I want to bake this a little bit before factoring out the client. I think the code duplication is ok for now as being extremely explicit is not a bad thing in these tests. Let me know if you feel strongly though, happy to discuss further.

Reply:
Can always discuss further or address later.